src/shogun/machine/gp/SparseVGInferenceMethod.cpp

```diff
@@ -0,0 +1,436 @@
+
+	//Tnm = Tnm + (model.y*yKnmInvA); 
+	eigen_Tnm += (eigen_y-eigen_m)*eigen_alpha.transpose();
+}
+
+SGVector<float64_t> CSparseVGInferenceMethod::get_posterior_mean()
+{
+	SG_NOTIMPLEMENTED
+	//TODO: implement this method once I get time
+	return SGVector<float64_t>();
+}
+
+SGMatrix<float64_t> CSparseVGInferenceMethod::get_posterior_covariance()
+{
+	SG_NOTIMPLEMENTED
+	//TODO: implement this method once I get time
+	return SGMatrix<float64_t>();
+}
+
+SGVector<float64_t> CSparseVGInferenceMethod::get_derivative_wrt_likelihood_model(
+		const TParameter* param)
+{
+	REQUIRE(!strcmp(param->m_name, "log_sigma"), "Can't compute derivative of "
+			"the nagative log marginal likelihood wrt %s.%s parameter\n",
+			m_model->get_name(), param->m_name)
+
+	SGVector<float64_t> dlik(1);
+
+	Map<VectorXd> eigen_alpha(m_alpha.vector, m_alpha.vlen);
+	Map<MatrixXd> eigen_inv_La(m_inv_La.matrix, m_inv_La.num_rows, m_inv_La.num_cols);
+	Map<MatrixXd> eigen_ktru(m_ktru.matrix, m_ktru.num_rows, m_ktru.num_cols);
+	Map<MatrixXd> eigen_kuu(m_kuu.matrix, m_kuu.num_rows, m_kuu.num_cols);
+	//yKnmInvLmInvLainvLa = yKnmInvLmInvLa*invLa';
+	//sigma2aux = sigma2*sum(sum(invLa.*invLa))  + yKnmInvLmInvLainvLa*yKnmInvLmInvLainvLa';
+	float64_t sigma2aux= m_sigma2*eigen_inv_La.cwiseProduct(eigen_inv_La).array().sum()+
+		eigen_alpha.transpose()*(eigen_kuu*CMath::exp(m_log_scale*2.0)
+			+CMath::exp(m_log_ind_noise)*MatrixXd::Identity(
+		m_kuu.num_rows, m_kuu.num_cols))*eigen_alpha;
+	//Dlik_neg = - (model.n-model.m) + model.yy/sigma2 - 2*F3 - sigma2aux - 2*TrK;
+	
+	dlik[0]=(m_ktru.num_cols-m_ktru.num_rows)-m_yy/m_sigma2+2.0*m_f3+sigma2aux+2.0*m_trk;
+	return dlik;
+}
+
+SGVector<float64_t> CSparseVGInferenceMethod::get_derivative_wrt_inducing_features(
+	const TParameter* param)
+{
+	//[DXu DXunm] = kernelSparseGradInd(model, Tmm, Tnm);
+    //DXu_neg = DXu + DXunm/model.sigma2;
+	
+	Map<MatrixXd> eigen_Tmm(m_Tmm.matrix, m_Tmm.num_rows, m_Tmm.num_cols);
+	Map<MatrixXd> eigen_Tnm(m_Tnm.matrix, m_Tnm.num_rows, m_Tnm.num_cols);
+
+	int32_t dim=m_inducing_features.num_rows;
+	int32_t num_samples=m_inducing_features.num_cols;
+	SGVector<float64_t>deriv_lat(dim*num_samples);
+	deriv_lat.zero();
+
+	m_lock->lock();
+	CFeatures *inducing_features=get_inducing_features();
+	//asymtric part (related to xu and x)
+	m_kernel->init(inducing_features, m_features);
+	for(int32_t lat_idx=0; lat_idx<eigen_Tnm.cols(); lat_idx++)
+	{
+		Map<VectorXd> deriv_lat_col_vec(deriv_lat.vector+lat_idx*dim,dim);
+		//p by n
+		SGMatrix<float64_t> deriv_mat=m_kernel->get_parameter_gradient(param, lat_idx);
+		Map<MatrixXd> eigen_deriv_mat(deriv_mat.matrix, deriv_mat.num_rows, deriv_mat.num_cols);
+		//DXunm/model.sigma2;
+		deriv_lat_col_vec+=eigen_deriv_mat*(-CMath::exp(m_log_scale*2.0)/m_sigma2*eigen_Tnm.col(lat_idx));
+	}
+
+	//symtric part (related to xu and xu)
+	m_kernel->init(inducing_features, inducing_features);
+	for(int32_t lat_lidx=0; lat_lidx<eigen_Tmm.cols(); lat_lidx++)
+	{
```

Comment:
Does second loop depend on first? if not one could replace by a single one?
Parallelise? Or doesnt make sense?